sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/UnpredictableHashSaltCheck.kt

```diff
@@ -0,0 +1,150 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.kotlin.checks
+
+import org.jetbrains.kotlin.psi.KtBlockExpression
+import org.jetbrains.kotlin.psi.KtCallExpression
+import org.jetbrains.kotlin.psi.KtExpression
+import org.jetbrains.kotlin.psi.KtNameReferenceExpression
+import org.jetbrains.kotlin.psi.psiUtil.collectDescendantsOfType
+import org.jetbrains.kotlin.psi.psiUtil.getParentOfType
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.calls.callUtil.getCall
+import org.jetbrains.kotlin.resolve.calls.callUtil.getResolvedCall
+import org.jetbrains.kotlin.resolve.calls.model.ResolvedCall
+import org.jetbrains.kotlin.resolve.scopes.receivers.ExpressionReceiver
+import org.sonar.check.Rule
+import org.sonarsource.kotlin.api.CallAbstractCheckWithData
+import org.sonarsource.kotlin.api.ConstructorMatcher
+import org.sonarsource.kotlin.api.FunMatcher
+import org.sonarsource.kotlin.api.getCalleeOrUnwrappedGetMethod
+import org.sonarsource.kotlin.api.matches
+import org.sonarsource.kotlin.api.predictRuntimeIntValue
+import org.sonarsource.kotlin.api.predictRuntimeStringValue
+import org.sonarsource.kotlin.api.predictRuntimeValueExpression
+import org.sonarsource.kotlin.api.secondaryOf
+import org.sonarsource.kotlin.plugin.KotlinFileContext
+
+private const val SPECS_PACKAGE = "javax.crypto.spec"
+private const val KEY_SPEC_FUN_NAME = "PBEKeySpec"
+private const val PARAMETER_SPEC_FUN_NAME = "PBEParameterSpec"
+private val STRING_TO_BYTE_FUNS = listOf(
+    FunMatcher(qualifier = "kotlin.text", name = "toByteArray"),
+    FunMatcher(qualifier = "java.lang.String", name = "getBytes")
+)
+private val BYTE_ARRAY_CONSTRUCTOR = ConstructorMatcher("kotlin.ByteArray")
+private val BYTE_ARRAY_CONSTRUCTOR_SIZE_ARG_ONLY = ConstructorMatcher("kotlin.ByteArray") { withArguments("kotlin.Int") }
+private val SECURE_RANDOM_FUNS = FunMatcher(qualifier = "java.security.SecureRandom")
+
+private const val MSG_MAKE_UNPREDICTABLE = "Make this salt unpredictable."
+private const val MSG_ADD_SALT = "Add an unpredictable salt value to this hash."
+private const val MSG_MIN_LEN = "Make this salt at least 16 bytes."
+private const val SMSG_PREDICTABLE_SALT = "Predictable salt value"
+
+@Rule(key = "S2053")
+class UnpredictableHashSaltCheck : CallAbstractCheckWithData<Int>() {
+    override val functionsToVisit = mapOf(
+        ConstructorMatcher("$SPECS_PACKAGE.$KEY_SPEC_FUN_NAME") to 1,
```

Comment:
I introduced this because I've seen this pattern (having to decide something based on which exact FunMatcher has matched) multiple times now. It doesn't seem very elegant and more bug-prone to me to have to check the name every time. It introduces redundant checks and you'll always need to maintain both in parallel when you touch the rule - which you could forget more easily. This way I thought was a relatively simple way to deal with those situations and make a decision on certain things already when matching. (Well and also it reduces the number of string comparisons we do - not that that will make much of a difference, most likely). My idea was to simplify rules, not make them more complicated, by taking out this generic 'boilerplate' logic and extracting it into `CallAbstractCheckWithData`. How do you feel does it make the implementation more complicated?

Reply:
I got your point and yes sometimes we have too. Practically, String comparison is not a big problem here but if we want to solve it nicer we can use your approach and maybe modify it a little.

If the problem is just to know which Matcher was actually matched, how about adding the Matcher itself as a parameter (instead of a generic data) or maybe it's index in the list of matchers and do the change in CallAbstractMatcher instead of creating the new class?